frappe/utils/excel.py

```diff
@@ -0,0 +1,238 @@
+
+	return current_row
+
+def print_value(df, doc, meta, parent_doc=None):
+	if df.fieldtype=="Check":
+		# print 'Checkbox'
+		return df.fieldname
+	elif df.fieldtype=="Image":
+		# print 'Image'
+		return doc[doc.meta.get_field(df.fieldname).options]
+	elif df.fieldtype in ("Attach", "Attach Image") and doc[df.fieldname] and (guess_mimetype(doc[df.fieldname])[0] or "").startswith("image/"):
+		# print 'Attach Image'
+		doc[df.fieldname]
+	elif df.fieldtype=="HTML":
+		# print 'HTML'
+		return "some html -- more processing"
+	else:
+		# print 'something else'
+		return doc.get_formatted(df.fieldname, doc, translated=True)
+
+
+def get_visible_columns(data, table_meta, df):
+	"""Returns list of visible columns based on print_hide and if all columns have value."""
+	columns = []
+	doc = data[0] or frappe.new_doc(df.options)
+	def add_column(col_df):
+		return is_visible(col_df, doc) \
+			and column_has_value(data, col_df.get("fieldname"))
+
+	if df.get("visible_columns"):
+		# columns specified by column builder
+		for col_df in df.get("visible_columns"):
+			# load default docfield properties
+			docfield = table_meta.get_field(col_df.get("fieldname"))
+			if not docfield:
+				continue
+			newdf = docfield.as_dict().copy()
+			newdf.update(col_df)
+			if add_column(newdf):
+				columns.append(newdf)
+	else:
+		for col_df in table_meta.fields:
+			if add_column(col_df):
+				columns.append(col_df)
+
+	return columns
+
+
+def column_has_value(data, fieldname):
+	"""Check if at least one cell in column has non-zero and non-blank value"""
+	has_value = False
+
+	for row in data:
+		value = row.get(fieldname)
+		if value:
+			if isinstance(value, basestring):
+				if strip_html(value).strip():
+					has_value = True
+					break
+			else:
+				has_value = True
+				break
+
+	return has_value
+
+
+def validate_print_permission(doc):
+	if frappe.form_dict.get("key"):
+		if frappe.form_dict.key == doc.get_signature():
+			return
+
+	for ptype in ("read", "print"):
+		if (not frappe.has_permission(doc.doctype, ptype, doc)
+			and not frappe.has_website_permission(doc.doctype, ptype, doc)):
+			raise frappe.PermissionError(_("No {0} permission").format(ptype))
+
+
+def has_value(df, doc):
+	value = doc.get(df.fieldname)
+	if value in (None, ""):
+		return False
+
+	elif isinstance(value, basestring) and not strip_html(value).strip():
+		return False
+
+	elif isinstance(value, list) and not len(value):
+		return False
+
+	return True
+
+
+def is_visible(df, doc):
+	"""Returns True if docfield is visible in print layout and does not have print_hide set."""
+	if df.fieldtype in ("Section Break", "Column Break", "Button"):
+		return False
+
+	if hasattr(doc, "hide_in_print_layout"):
+		if df.fieldname in doc.hide_in_print_layout:
+			return False
+
+	if df.permlevel > 0 and not doc.has_permlevel_access_to(df.fieldname, df):
+		return False
+
+	return not doc.is_print_hide(df.fieldname, df)
+
+
+def get_letter_head(doc, no_letterhead):
```

Comment:
Yes. Its very bad I agree. It was a quick fix since the client really wanted this feature. I tried importing the functions using 
`from frappe.utils.print import blah, blah, blah`
but the compiler didn't like the word print in the from statement. Any ideas